arbnode/espresso_state_checker.go

```diff
@@ -0,0 +1,100 @@
+package arbnode
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+	flag "github.com/spf13/pflag"
+)
+
+type StateCheckerConfig struct {
+	Enable          bool          `koanf:"enable"`
+	PollingInterval time.Duration `koanf:"polling-interval"`
+
+	// http endpoint of the trusted node
+	TrustedNodeUrl string `koanf:"trusted-node-url"`
+}
+
+var DefaultStateCheckerConfig = StateCheckerConfig{
+	Enable:          true,
+	PollingInterval: time.Second * 100,
+}
+
+func EspressoStateCheckerConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultStateCheckerConfig.Enable, "enable state checker")
+	f.Duration(prefix+".polling-interval", DefaultStateCheckerConfig.PollingInterval, "time after a success")
+	f.String(prefix+".trusted-node-url", DefaultStateCheckerConfig.TrustedNodeUrl, "http endpoint of the trusted node")
+}
+
+type StateChecker struct {
+	stopwaiter.StopWaiter
+
+	config       StateCheckerConfig
+	fatalErrChan chan error
+
+	trustedClient *ethclient.Client
+	myClient      *ethclient.Client
+}
+
+func NewStateChecker(
+	config StateCheckerConfig,
+	httpPort int,
+	fatalErrChan chan error,
+) *StateChecker {
+	if config.TrustedNodeUrl == "" {
+		return nil
+	}
+
+	client, err := ethclient.DialContext(context.Background(), config.TrustedNodeUrl)
+	if err != nil {
+		panic(err)
+	}
+	myUrl := fmt.Sprintf("http://localhost:%d", httpPort)
```

Comment:
will localhost always work?

Reply:
`state checker` is a component working inside the caff node. So we should make sure `localhost` is working